src/Symfony/Component/Notifier/Message/PushMessage.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier\Message;
+
+use Symfony\Component\Notifier\Exception\InvalidArgumentException;
+use Symfony\Component\Notifier\Notification\Notification;
+use Symfony\Component\Notifier\Recipient\PushRecipientInterface;
+
+/**
+ * @author Tomas Norkūnas <norkunas.tom@gmail.com>
+ */
+final class PushMessage implements MessageInterface
+{
+    private $transport;
+    private $recipientId;
+    private $subject;
+    private $content;
+    private $options;
+    private $notification;
+
+    public function __construct(string $recipientId, string $subject, string $content, MessageOptionsInterface $options = null)
```

Comment:
Some services, like [Pushbullet](https://docs.pushbullet.com/#create-push), don't require an explicit recipient<sup>*</sup> because they are designed to send push notifications to a single user (provided by an OAuth2 token in the HTTP header).  How should we support those types of services?  Furthermore...

Should that OAuth2 token be provided in the DSN? That would follow the same pattern that other Notifier integrations use (like Slack).  And it would easily support use cases where a single website admin is the only person who will receive those notifications.  In that case, requiring a `$recipientId` here doesn't make sense, **so perhaps we should make that constructor argument nullable** given that `MessageInterface::getReceipientId()` supports `null` values?

Should that OAuth2 token be provided via this `$recipientId` argument instead?  It would support more advanced use cases, like websites where the users can each connect their individual accounts to Pushbullet - in those situations, providing an access key in a DSN doesn't make much sense.

Or perhaps we need to support both approaches for services like Pushbullet.  In other words, when sending a push notification with Pushbullet, we'd first look at the `$recipientId` to see if one exists; if so, we'd use that, otherwise we'd fall back to an (optional) access key from the DSN, or throw an exception if neither exists.

I know Pushbullet is being implemented here, but I have started working on that integration using this code, and I just want to ensure that this new `PushMessage` class will support future services like Pushbullet.

---

<sup>*</sup> They do support four optional parameters for sending the message to a particular device, user, channel, or client, but none of those are required, and two of them take identical format, making it impossible to leverage a single `$recipientId` string - those would be better implemented using custom options.

Reply:
At last someone is interested :) As I'm not familiar with Pushbullet so I didn't thought about this. If provider has a multiple ways to set recipients then I think it's okay to make that arg as nullable as OneSignal also supports sending via multiple ways like Pushbullet.. But the difference is here that OneSignal requires at least 1 filter where to send that notification while Pushbullet not? If yes,d o you have a suggestion how transport should behave if no recipients are provided? :)

Reply:
> At last someone is interested :)

I am interested as well ❤️ 😃

Reply:
> But the difference is here that OneSignal requires at least 1 filter where to send that notification while Pushbullet not? If yes,d o you have a suggestion how transport should behave if no recipients are provided? :)

Yes, you are correct, and I have two suggestions :)

1. Your transport's `supports()` method should return `false` if `$message->getRecipientId() === null`
2. Your transport's `doSend()` method should also check if that field is `null` and throw an exception

Reply:
Thanks will do this way then, just if someone will need to send by other filters OneSignal notifications it will need refactor :)
But for the first implementation I think single recipient is enough.

Reply:
On a related note, in https://github.com/symfony/symfony/pull/35167#issuecomment-571929527 the mandatory `$recipient` argument [was removed](https://github.com/symfony/symfony/pull/35167/files#diff-d7862467203ee6fa24a49bfd01b6a789cff844585f4fec58caaa878e0544bf39) from `ChatMessage::fromNotification()`:

https://github.com/symfony/symfony/blob/c71c8727cc665c9e9b56e299fcfcc0adfbf02bac/src/Symfony/Component/Notifier/Message/ChatMessage.php#L32-L38

Perhaps we should follow that same pattern to allow the creation of messages without a recipient? See line 32:

https://github.com/symfony/symfony/blob/c71c8727cc665c9e9b56e299fcfcc0adfbf02bac/src/Symfony/Component/Notifier/Channel/ChatChannel.php#L24-L33

WDYT?

Reply:
yes, already in progress :)

Reply:
Updated :)